hazelcast-hibernate/hazelcast-hibernate4/src/main/java/com/hazelcast/hibernate/local/LocalRegionCache.java

```diff
@@ -195,18 +193,28 @@ protected Object createMessage(final Object key, final Object value, final Objec
     protected MessageListener<Object> createMessageListener() {
         return new MessageListener<Object>() {
             public void onMessage(final Message<Object> message) {
+                if (message.getPublishingMember().localMember()) {
+                    return;
+                }
                 final Invalidation invalidation = (Invalidation) message.getMessageObject();
-                if (versionComparator != null) {
-                    final Expirable value = cache.get(invalidation.getKey());
-                    if (value != null) {
-                        Object currentVersion = value.getVersion();
-                        Object newVersion = invalidation.getVersion();
-                        if (versionComparator.compare(newVersion, currentVersion) > 0) {
-                            cache.remove(invalidation.getKey(), value);
+                Object key = invalidation.getKey();
+                if (key != null) {
+                    if (versionComparator != null) {
+                        final Expirable value = cache.get(key);
+                        if (value != null) {
+                            Object newVersion = invalidation.getVersion();
+                            if (newVersion != null) {
+                                Object currentVersion = value.getVersion();
+                                if (versionComparator.compare(newVersion, currentVersion) > 0) {
+                                    cache.remove(key, value);
+                                    return;
+                                }
+                            }
                         }
                     }
+                    cache.remove(key);
```

Comment:
You are right, I put the `return` in the wrong place.  Not only this, now that I actually look hard at the version comparison, it's actually wrong.  I've just reworked this logic and put it into its own `maybeInvalidate` method as you suggested.  My comments in there try to explain the logic.